python/python/tests/test_async_client.py

```diff
@@ -7951,6 +7953,125 @@ async def test_function_delete_with_routing(
+                    result = await redis_client.function_kill()
+                    #  we expect to get success
+                    assert result == "OK"
+                    break
+                except RequestError:
+                    # a RequestError may occur if the function is not yet running
+                    # sleep and try again
+                    timeout += 0.5
+                    await asyncio.sleep(0.5)
+
+        await asyncio.gather(
+            endless_fcall_route_call(),
+            wait_and_function_kill(),
+        )
+
+        # no functions running so we get notbusy error again
+        with pytest.raises(RequestError) as e:
+            assert await redis_client.function_kill()
+        assert "NotBusy" in str(e)
+
+    @pytest.mark.parametrize("cluster_mode", [False, True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_function_kill_write_is_unkillable(
+        self, request, cluster_mode, protocol, redis_client: TGlideClient
+    ):
+        min_version = "7.0.0"
+        if await check_if_server_version_lt(redis_client, min_version):
+            return pytest.mark.skip(reason=f"Redis version required >= {min_version}")
+
+        lib_name = f"mylib1C{get_random_string(5)}"
+        func_name = f"myfunc1c{get_random_string(5)}"
+        code = create_lua_lib_with_long_running_function(lib_name, func_name, 10, False)
+
+        # load the library on all primaries
+        assert await redis_client.function_load(code, replace=True) == lib_name.encode()
+
+        # create a second client to run fcall - and give it a long timeout
+        test_client = await create_client(
+            request, cluster_mode=cluster_mode, protocol=protocol, timeout=15000
+        )
+
+        # call fcall to run the function loaded function
+        async def endless_fcall_route_call():
+            # fcall won't be killed, because kill only works against fcalls that don't make a write operation
+            # use fcall(key) so that it makes a write operation
+            await test_client.fcall(func_name, keys=[lib_name])
+
+        async def wait_and_function_kill():
+            # it can take a few seconds for FCALL to register as running
+            await asyncio.sleep(3)
+            timeout = 0
+            foundUnkillable = False
+            while timeout <= 5:
+                # keep trying to kill until we get a unkillable return error
+                try:
+                    await redis_client.function_kill()
+                except RequestError as e:
+                    if "UNKILLABLE" in str(e):
+                        foundUnkillable = True
+                        break
+                timeout += 0.5
+                await asyncio.sleep(0.5)
+            # expect an unkillable error
+            assert foundUnkillable
+
+        await asyncio.gather(
+            endless_fcall_route_call(),
+            wait_and_function_kill(),
+        )
```

Comment:
i see that the steps afterwards are gone, namely retest that no function is running that can be killed, were they causing flakeyness?

Reply:
the function stats calls were causing all sorts of flakiness.  We don't actually need to do a function stat, and I figured this direction would be more stable.

Reply:
Do you want to remove/update this piece from java client then? (maybe in another PR)